apps/src/blockly/addons/cdoBlockShadow.js

```diff
@@ -0,0 +1,141 @@
+            // If tthis block's workspace id matches the flyout workspace id,
+            // then we have found the parent.
+            if (flyoutWorkspace && flyoutWorkspace.id === block.workspace.id) {
+              parent = potentialParent;
+            }
+          }
+        }
+      }
+    }
+  }
+  if (parent) {
+    return getImageUrlFromParent(parent, blockShadowData.imageIndex);
+  } else {
+    // The block is probably disconnected from any root or toolbox. Reset to
+    // default text.
+    return '';
+  }
+}
+
+// Given a shadow block and an optional parent block id, update the image displayed
+// on the shadow block to match the image selected in the parent block. See getShadowedBlockImageUrl
+// for details on how we find the parent block.
+export function updateShadowedBlockImage(block, parentId) {
+  const url = getShadowedBlockImageUrl(block, parentId);
+  changeShadowedImage(url, block);
+}
+
+// Get the image url from the parent block's inputList at the given index.
+// We find the image url by looking at the connection to the input at the given index.
+// If the parent does not have an image at the given index, return an empty string.
+function getImageUrlFromParent(parentBlock, imageIndexOnParent) {
+  const targetConnection =
+    parentBlock.inputList[imageIndexOnParent]?.connection?.targetConnection;
+  // We only want to get the image from a connection block that is not an insertion marker.
+  // If the block is an insertion marker that means the block is being dragged
+  // and is not yet connected to the parent block.
+  if (
+    targetConnection &&
+    targetConnection.sourceBlock_ &&
+    !targetConnection.sourceBlock_.isInsertionMarker()
+  ) {
+    return (
+      targetConnection.sourceBlock_.inputList[0]?.fieldRow[0]?.imageElement_?.getAttribute(
+        'xlink:href'
+      ) || ''
+    );
+  } else {
+    return '';
+  }
+}
+
+// Set the given childBlock's image input to the given imageUrl,
+// or reset the image input to the default long text if the imageUrl is empty.
+function changeShadowedImage(imageUrl, childBlock) {
+  if (!imageUrl || imageUrl.length === 0) {
+    resetShadowedImageToLongString(childBlock);
+    return;
+  }
+  const textInput = childBlock.inputList[0].fieldRow[0];
+  const previewInput = childBlock.inputList[0].fieldRow[1];
+  textInput.setValue(childBlock.shortString);
+  previewInput.setValue(imageUrl);
+  previewInput.updateDimensions(
+    childBlock.thumbnailSize,
+    childBlock.thumbnailSize
+  );
+  previewInput.getSize();
+}
+
+// Reset the child block to no longer have an image input, and only
+// show the long text input.
+function resetShadowedImageToLongString(childBlock) {
+  const textInput = childBlock.inputList[0].fieldRow[0];
+  const previewInput = childBlock.inputList[0].fieldRow[1];
+  textInput.setValue(childBlock.longString);
+  previewInput.setValue('');
+  previewInput.updateDimensions(1, childBlock.thumbnailSize);
```

Comment:
Does `1` mean a width of 1 pixel? I'm curious why we're using the existing size for height. Would it make sense to use a variable (or comment) to explain the number literal?

Reply:
yeah 1 means 1 pixel. I just kept the existing functionality here, I can see what happens if we use 0 for example.

Reply:
I tested this out. We can't have a width or height of 0. For some reason if we don't set the height from the start the image won't show up in the toolbox, not quite sure why. I'll add a comment.

Reply:
Yeah it seems like the height will not change even if we manually try to change it later, but the width will. Very odd!

Reply:
Interesting - maybe something in how the original class is implemented? In any case I don't think it's a big deal as long as we have a comment to explain. Thanks for looking into it!